# Laboratory excercise 5

The notebook contains excercise connected to the auditory excerice 6. For any questions feel free to contact assistant: eda.jovicic@fer.hr

In [1]:
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf

# set seed for consistency
np.random.seed(2)

1. Load the cleaned dataset from the first exercise. If you haven't saved the dataset, rerun the exercise and save the final dataset.

In [2]:
file_path = r"C:\Users\ONE\OneDrive - FFUNIZG\Radna površina\Faks\CogSci\DataScience\data\student_data1.csv"
df = pd.read_csv(file_path)

2. Using linear regression, model the MathScore using features: TestPrep, NrSiblings, ParentMaritalStatus and ReadingScore. 

In [4]:
#Standardization procedure. TestPrep and ParentMaritalStatus are excluded since they are Boolean, while NrSiblings
#standardization doesn't make sense due to its structure and meaning.

# variables = ["MathScore","ReadingScore"]
# for var in variables:
#     df[var] = (df[var]-df[var].mean())/df[var].std()

In [5]:
model1 = smf.ols(formula="MathScore ~ C(TestPrep) + NrSiblings + C(ParentMaritalStatus) + ReadingScore", data=df)

res1 = model1.fit()
res1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              MathScore   R-squared:                       0.670
Model:                            OLS   Adj. R-squared:                  0.669
Method:                 Least Squares   F-statistic:                     9829.
Date:                Tue, 09 Jan 2024   Prob (F-statistic):               0.00
Time:                        23:43:25   Log-Likelihood:            -1.0471e+05
No. Observations:               29117   AIC:                         2.094e+05
Df Residuals:                   29110   BIC:                         2.095e+05
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                       6.2302      0.314     19.867      0.000       5.616       6.845
C(TestPrep)[T.1]                1.1060      0.113      9.807      0.000       0.885       1.327
C(ParentMaritalStatus)[T.1]     0.1733      0.145      1.193      0.233      -0.112       0.458
C(ParentMaritalStatus)[T.2]    -0.1318      0.168     -0.786      0.432      -0.460       0.197
C(ParentMaritalStatus)[T.3]     0.5432      0.396      1.373      0.170      -0.232       1.319
NrSiblings                      0.0155      0.036      0.428      0.668      -0.055       0.086
ReadingScore                    0.8570      0.004    239.510      0.000       0.850       0.864
==============================================================================
Omnibus:                      233.486   Durbin-Watson:                   1.987
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              155.618
Skew:                          -0.010   Prob(JB):                     1.61e-34
Kurtosis:                       2.642   Cond. No.                         562.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [3]:
#A model with interactions included.
#model2 = smf.ols(formula="MathScore ~ C(TestPrep)*C(ParentMaritalStatus)*ReadingScore*NrSiblings", data=df)

#res2 = model.fit()
#res2.summary()

Questions:

a) Using the summary, write down the formula with coefficients

$MathScore = 6.2302 + 0.8570×ReadingScore + 1.106 * TestPrep + (0.1733*ParentMaritalStatus[T.1] −0.1318×C(ParentMaritalStatus)[T.2] + 0.5432×C(ParentMaritalStatus)[T.3]) + 0.0155×NrSiblings$

ParentMaritalStatus is added depending on the category the row belongs to: if it belongs to class 0, none of the elements are added, while if it belongs to 1,2, or 3 only the corresponding ParentMaritalStatus [T1], [T2] or [T3] is scaled and added while others are ignored.

b) Are there any statisticaly important features?

Yes, for the *p<.05* level of significance:
**TestPrep** (*p<.001*) and **ReadingScore** (*p<.001*) are statisctically important features.

c) Let's interpret the data, if students do test preparation, how does that affect the outcome on the math exam? 

For the given model, if a student does test preparation (**TestPrep [T.1]**), leaving all other variable values unchanged, the outcome of the math exam is boosted by 1.106 points.

d) Are students who practice sport better at math exams or worse? Using linear regression, model the MathScore using only one feature: PracticeSport. Don't include intercept. Using summary, decide which students do the best at math exams - those who never practice sport, sometimes practice sport or those who do it regularly?

In [6]:
# Declare the model
mod_sp = smf.ols(formula = "MathScore ~ C(PracticeSport)", data=df)
res_sp = mod_sp.fit()
res_sp.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              MathScore   R-squared:                       0.006
Model:                            OLS   Adj. R-squared:                  0.006
Method:                 Least Squares   F-statistic:                     87.35
Date:                Tue, 09 Jan 2024   Prob (F-statistic):           1.51e-38
Time:                        23:51:19   Log-Likelihood:            -1.2074e+05
No. Observations:               29117   AIC:                         2.415e+05
Df Residuals:                   29114   BIC:                         2.415e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=========================================================================================
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
Intercept                64.1576      0.244    262.426      0.000      63.678      64.637
C(PracticeSport)[T.1]     3.6865      0.287     12.858      0.000       3.125       4.248
C(PracticeSport)[T.2]     2.1320      0.275      7.753      0.000       1.593       2.671
==============================================================================
Omnibus:                      234.462   Durbin-Watson:                   1.981
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              214.722
Skew:                          -0.171   Prob(JB):                     2.36e-47
Kurtosis:                       2.756   Cond. No.                         5.81
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

(**PracticeSport** label encoding: 0 signifies *Never*, 1 signifies *regularly*, 2 signifies *sometimes*.)

Those who *regularly* (**T1**) practise sport perform best at the math exam - with +3.6865 in comparison to those *never* practising sport; while those practising sport *sometimes* (**T2**) perform second best - with +2.1320 results with regards to those *never* practising.